### PR TITLE
weighted_moments

### LINKED ISSUE #15974: weighted_moments

### PR DESCRIPTION
**Pull request by usayrimrehman**
- Close  #15974

### CODE REVIEW

**Comment by usayrimrehman** [comment]
@Ookamice 
Hey, I've tried incorporating the changes you asked for. Can you give them a look, and let me if they solve the problem.

**Comment by Ookamice** [comment]
@usayrimrehman
Hiya, just ran your tests in the codespace and there seems to be an error that is produced with `sum()`. I'm not 100% sure where the problem is exactly but if you have any thoughts, feel free to share it and we can work together in solving this.

Let me know if you have had a closer look at the error logs and if the tests are runnable working on your end. 😄

**Comment by usayrimrehman** [comment]
@Ookamice
Hey, I am having trouble setting up pytests both locally, and on the Github codespace. I tried to follow the steps mentioned in the docs, but can't seem to fix it.
I'd really appreciate it, if you can spare some time, and guide me, either here or by setting up a meeting depending on what you're comfortable with.

Thank you so much.

**Comment by Ookamice** [comment]
@usayrimrehman
Let me know which part of https://unify.ai/docs/ivy/overview/contributing/setting_up.html#github-codespaces you're struggling with. Vitally, please note that you click New with options
image

**Comment by usayrimrehman** [comment]
@Ookamice 
Thank you so much for helping me set up the tests. I gave the issues you pointed out a look, and have managed to get past them. The tests, however, are still failing, and the error message I get is the return, and ground truth dimensions don't match (see image1 below). Upon checking the error message in detail ( check image 2), it can be seen that the returned, and ground truth values are the same, and it is indeed only the shapes that don't match.  I can't seem to get around why the shapes don't align, and would be really grateful if you help me out a little.

Image 1
image

Image 2
image

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by usayrimrehman** [comment]
I'm still working on the issue, and will commit as soon as I am done.

**Comment by usayrimrehman** [comment]
@Ookamice 
I am really sorry for disturbing you again, but I genuinely need help with the issue I am stuck at. First of all, I've using the TensorFlow documentation you suggested from the start, and the ivy front frontend function I wrote is closely based on the original TensorFlow function.

I am gonna start by removing the last two lines, which squeeze the output when keepdims = False. Upon testing the function  after this I get the following error message.
image
The following image shows that the returned, and the ground-truth values are the same, and it's the shapes that don't match because the returned shape is (1,), while the ground-truth shape is ().
image
To solve this temporarily, I wrote the following two lines of code:
```
weighted_mean=weighted_mean[0]
weighted_variance=weighted_variance[0] 
```
I was expecting this to solve the problem by changing the returned shape to (). However, I now received the following error message:  
image
The returned shape was now indeed (); however, the ground-truth shape, contrary to () previously, now came out to be (1,). You can again, in the following image, see that the values still add up, and it's the shapes that don't match.
image
As another gimmicky workaround for the time being, I wrote the following lines of code at the end, to see change the returned shape from (1,) to (), and vice versa:
```
if weighted_mean.shape == ():
        weighted_mean=ivy.array([weighted_mean])
    elif weighted_mean.shape == (1,):
        weighted_mean=weighted_mean[0]
    if weighted_variance.shape == ():
        weighted_variance=ivy.array([weighted_variance])
    elif weighted_variance.shape == (1,):
        weighted_variance=weighted_variance[0]   
```
However, the issue, as evident in the following images, still persists:
image
image
I am really confused as to why this is happening, and why changing the output shape impacts the ground-truth shape. I've reading Ivy's documentation and code on testing in detail, along with TensorFlow documentation on the subject, but haven't been able to find a solution. 
I would, therefore, be really grateful, if you could lend me a hand in fixing this issue, or at least point me in the right direction.

**Comment by usayrimrehman** [comment]
@Ookamice
Hey, sorry to disturb you again. Were you able to give the issue a look?
Thank you in advance.

**Comment by Ookamice** [comment]
@usayrimrehman 
Hiya, the ground-truth shape will depend on the input given to it. Due to the random nature of the automated tests, every test will have a slightly different input  and so this will impact the output of both your implementation and tensorflow's. (Without seeing the inputs of the tests, I can't draw any conclusions on why the shapes are differing). Identifying bugs from the automated tests is pretty difficult especially for larger functions like yours.

One suggestion I would make is trying to write some small minimum code examples on super simple test cases to see if your function deviates. It is always a good idea to do some basic sanity checks before trying to work out the bugs in edge-cases. Here is the following code snippets that may aid you:
```python
import ivy.functional.frontends.tensorflow as ivy_tf
import tensorflow as tf
# ...
print(tf.nn.weighted_moments(x,axes,frequency_weights))
print(ivy_tf.nn.weighted_moments(x,axes,frequency_weights))
```

If this does not help, a more tedious but thorough method would be stepping through your code one function at a time to see where along the chain of operations does your output start differing from what you would have expected.

**Comment by usayrimrehman** [comment]
@Ookamice
Hey, that is actually a really good suggestion. I'll try to work in this direction, and get back to you, if something bogs me down, again.
Thank you so much, again.

**Comment by usayrimrehman** [comment]
@Ookamice 
Hey!
So, I, due to suggestion, was able to figure out why the tests were failing, and was able to resolve them (to some extent at least). If you check the image below, all tests apart from the one with the "numpy backend" passed.
image
I decided to give the error a look, and found this message:
image
The message says that the function produces a '-inf' array element, when it should actually produce '-1.502'. The error also provides the inputs it used. 
I tried testing the function myself with the inputs in the error message to see if the -inf is actually produced, and the results can be seen in the image below: 
image
It can be seen that the output from the ivy function matches the output from the TensorFlow function perfectly, and produces no -inf results. Therefore, I don't really get what's throwing the test off, when evidently it should output the right values.

**Comment by Ookamice** [comment]
@usayrimrehman 
Hiya, great work you're doing!! 🚀 
Could you quickly try that example again but with `ivy.set_backend('numpy')` added? Usually in the case of every test but one passing, it is usually an error to do with the Ivy functional API rather than an error with your implementation although it is odd that numpy is the one failing as from my experience, it is the backend with the least bugs 😅

**Comment by usayrimrehman** [comment]
@Ookamice 
Hey, I'm sorry I didn't get a notification when you edited the last comment, so didn't know the PR was ready to merge. I just notice the edit made to the comment, and have resolved all the related conflicts. Please, give it a look, when you find the time.

**Comment by Ookamice** [comment]
🚀 🥳
